### ISSUE TITLE
Support for UAC elevation in `msys2_shell.cmd`

### ISSUE THREAD

**Issue by elieux** [msys2/msys2#71]
Investigate this script: https://sourceforge.net/p/msys2/mailman/attachment/3c7110f2-15e6-a833-72a8-a0fb7eafc691%40gmail.com/1/

From: [Vincent Belaïche](https://sourceforge.net/p/msys2/mailman/message/36000301/)

**Comment by vincentb1** [comment]
Hello @elieux &amp; al.,
Shall I make at once the the change of having the VBScript a temporary file erased before exit. Or would you like some prior reading of my proposal.
Just some comments on what I have done:

- I use `if {`&hellip;`}=={`&hellip;`}` for string comparison, rather than `if "x`&hellip;`"=="x`&hellip;`"`, because the string may contain itself a double quote. That is not the case here, `"x`&hellip;`"` would just have been fine, but I think that it is a good habit to use this `{`&hellip;`}` syntax.
- I write `goto labelname`, rather than `goto :labelname`. The colon-prefixed notation should be, IMHO, used only for procedures (ie `call :labelname arg1 arg2`&hellip;)
- I had to rename the label `EOF` to `exit`, because `goto :EOF` is reserved for procedure return.
- I use upper case for command options. Command names like `del`, `for` or `start` are insensitive to case, but I am not sure that you can assume that in the future options will remain case-insensitive too. So I think it is better to write `exit /B` than `exit /b`, I did not fix it, just did it this way for my added code.
- Unfortunately, I don't know anyway to get the return status of the `.ShellExecute`-ed process. I don't know how in the VBScript wait for its exit, get the return status, and pass it to the `Wscript.Quit`. I don't think that this is an issue, because you already do some `exit /b 0` after `start` w/o using `start /W` to wait for exit.

**Comment by vincentb1** [comment]
Hello @elieux & al.,
I made an update to posted script [msys2_shell.cmd.remove-this-fnext](https://sourceforge.net/p/msys2/mailman/attachment/9b2a2424-52c8-da86-195b-851a65de979d%40gmail.com/1/). To use it, you need to remove the filename extension `.remove-this-fnext` which I added only for the carrying mail not to be blocked by the list server.
The changes are as follows:

- Finally, I reverted to using `"x`&hellip;`"` rather than `{`&hellip;`}` for string comparison, because the latter construct fails in case of a string containing spaces, and it is not such a big issue to ensure that the string in the former construct is w/o double quotes.
- The VBscript is created every time as a temporary file, and deleted after use. So there is no longer any such thing as ensuring consistency between `msys2_shell.cmd` and the VBScript script via some version number.
- At the beginning of the DOS script, I explicitely ensure DOS command extension by calling `setlocal EnableExtensions`. I observed that sometimes they were not enabled by default :confused:.

I am going on passing the command line arguments via an envvar. So I need to make `endlocal` at some point of time, and therefore the environment of the `msys2_shell.cmd` calling process is write-accessed&hellip; Well, I don't think that this is a big issue. It is possible to do otherwise, and I have alternative code to convert within MSDos the command line argument list to VScript format &mdash; this amounts to replaceing all `"` by `""` in it &mdash; and write it into the script code. However, this is more complex to do.

**Comment by vincentb1** [comment]
BTW, I noticed that the `goto` labels are most fo the time written with a colon when passed to `goto`. I don't think that this is a good practice, the colon should be reserved for keywords like EOF. So we should write:
```DOS
goto there
:there
```
Instead of 
```DOS
goto :there
:there
```
For reference see [Microsoft page on goto](https://msdn.microsoft.com/en-us/library/cc722866.aspx), there on can read:

> The label value you specify on the goto command line must match a label in the batch program. The label within the batch program must be preceded by a colon.

Writing that the label _must be preceded by a colon_ equates to considering that the colon is not part of the label itself. Still on the same page, in the example given there aren't any colon either when the `end` label  is passed to the `goto` command, ie the full instruction is `goto end`, not `goto :end`, as quoted below:
```DOS
echo off
format a: /s
if not errorlevel 1 goto end
echo An error occurred during formatting.
:end
echo Successfully formatted the disk in drive A.
```
I just quoted Microsoft's example, even though it is silly because in case of error that would display both error and success !  like this:
```
An error occurred during formatting.
Successfully formatted the disk in drive A.
```

**Comment by vincentb1** [comment]
Hello,
I saw that there is now a `msys/msys2-launcher-git` module that contains separate launchers for MSYS, MINGW64 and MINGW32. It's a good idea. Anyway, I did one more update to the `msys2_shell.cmd` script,  
[as attached msys2_shell.cmd.zip](https://github.com/msys2/msys2/files/1396326/msys2_shell.cmd.zip)

Changes are as follows:

- `MSYS2_PATH_TYPE` is also tunnelled to the console launched via Shell.Application ActiveX, without this fix, PATH variable inheritance would not be correctly handled.
- Command line options can be of the `-`_xxx_ or `--`_xxx_ form equivalently
- with `goto` _label_, _label_ is never with a colon prefix.
